### PR TITLE
Fixes #39418 - Fix inherited synced content after parent hostgroup OS change

### PR DESCRIPTION
**Pull request by nofaralfasi**
#### What are the changes introduced in this pull request?

- Updated hostgroup kickstart repository correction logic in app/models/katello/concerns/hostgroup_extensions.rb to re-evaluate synced content when the hostgroup OS changes, not only when the current repo becomes invalid.
- Enhanced equivalent kickstart repository selection to prefer candidates matching the new OS release, then narrow by distribution variant/product as fallback heuristics.
- Added a regression test in test/models/concerns/hostgroup_extensions_test.rb that covers parent OS change behavior and verifies child hostgroup inheritance reflects the updated kickstart repository.

#### Considerations taken when implementing this change?

- The bug occurs because a previously selected kickstart repo can still pass a broad validity check after OS minor-version changes, leaving stale synced content in inherited child hostgroups.
- The fix is scoped to OS-change and mismatch-recovery paths to avoid unnecessary repo churn in unrelated updates.
- Selection prioritizes release alignment first (to match expected provisioning media for the updated OS), while preserving existing compatibility intent through variant/product fallback.
- Inheritance behavior is preserved: only parent value correction is needed for children with inherited synced content to resolve automatically.

#### What are the testing steps for this pull request?
See redmine issue https://projects.theforeman.org/issues/39418

## Release Notes

* **Bug Fixes**
  * Fixed kickstart repository inheritance for child hostgroups so parent-based selections no longer fall back to unintended options.
  * Enhanced equivalent kickstart repository matching when an operating system changes, with improved handling of generic distribution versions and release-name/release-hint scenarios.
  * Updated recalculation behavior to reconcile install-media/medium changes first and only adjust kickstart repository selections when the effective resolved result would differ.

* **Tests**
  * Added coverage for child inheritance edge cases, OS-change propagation, stale child values, and generic distribution release-hint selection.

### CODE REVIEW

File: app/helpers/katello/kickstart_repository_helper.rb [hunk 1]
```diff
@@ -30,6 +33,11 @@ def kickstart_repository_id(host, options = {})
       # if the kickstart repo id is set in the host use that
       return host_ks_repo_id if host_ks_repo_id.present?
 
+      # Child hostgroups should remain on true inheritance unless explicitly overridden.
+      if host.is_a?(::Hostgroup) && host.parent_id.present?
```
Comment (lfu) [inline app/helpers/katello/kickstart_repository_helper.rb]: Could this be called earlier in the method?

File: app/models/katello/concerns/hostgroup_extensions.rb [hunk 2]
```diff
@@ -129,24 +123,158 @@ def add_organization_for_environment
...
+        equivalent_kickstart_repository_for_variant(ks_repos)
       end
 
       def matching_kickstart_repository?(content_facet)
         return true unless operatingsystem
 
         if operatingsystem.respond_to? :kickstart_repos
-          operatingsystem.kickstart_repos(self, content_facet: content_facet).any? do |repo|
+          effective_content_facet = effective_content_facet_for_kickstart(content_facet)
+          operatingsystem.kickstart_repos(self, content_facet: effective_content_facet).any? do |repo|
```
Comment (lfu) [inline app/models/katello/concerns/hostgroup_extensions.rb]: Issue: `operatingsystem.kickstart_repos()` is called twice with identical parameters during `correct_kickstart_repository`:

First call (line 151) - via `should_recalculate_kickstart_repository?` → `matching_kickstart_repository?(content_facet)` →   `operatingsystem.kickstart_repos(self, content_facet: effective_content_facet)`

Second call (line 128) - in `equivalent_kickstart_repository`:
  `operatingsystem.kickstart_repos(self, content_facet: effective_content_facet_for_kickstart(content_facet))`

Both calls use the same `hostgroup` (self) and same effective content facet, so the results should be identical.

Performance Impact: Medium - `kickstart_repos()` performs database queries to fetch available repositories. This happens on every `hostgroup` save when a kickstart repository exists.

Suggestion: Cache the result in `correct_kickstart_repository` and pass it to both methods:

```
def correct_kickstart_repository
  reconcile_medium_and_kickstart_repository
  
  ks_repos = cached_kickstart_repos  # fetch once
  return unless should_recalculate_kickstart_repository?(ks_repos)
  
  equivalent = equivalent_kickstart_repository(ks_repos)
  # ...
end
```
Then update method signatures to accept optional `ks_repos` parameter.

File: app/models/katello/concerns/hostgroup_extensions.rb [hunk 3]
```diff
@@ -129,24 +123,158 @@ def add_organization_for_environment
...
             repo[:id] == (content_facet&.kickstart_repository_id || content_facet&.kickstart_repository&.id)
           end
         end
       end
 
       private
 
+      def reconcile_medium_and_kickstart_repository
```
Comment (lfu) [inline app/models/katello/concerns/hostgroup_extensions.rb]: I was puzzled about these new method names.
How about `ensure_single_installation_source`?

File: app/models/katello/concerns/hostgroup_extensions.rb [hunk 4]
```diff
@@ -129,24 +123,158 @@ def add_organization_for_environment
...
+          # since it's :through association, nullify both the actual data source and delegate
+          self.content_facet.kickstart_repository = nil
+          self.kickstart_repository = nil
+        # If switched from install media to ks repo:
+        elsif content_facet&.kickstart_repository && medium
+          self.medium = nil
+        end
+      end
+
+      def should_recalculate_kickstart_repository?
```
Comment (lfu) [inline app/models/katello/concerns/hostgroup_extensions.rb]: How about `kickstart_repository_stale?`?

File: app/models/katello/concerns/hostgroup_extensions.rb [hunk 5]
```diff
@@ -129,24 +123,158 @@ def add_organization_for_environment
...
+      end
+
+      def operatingsystem_changed_for_recalculation?
+        changed_by_legacy_api = respond_to?(:operatingsystem_id_changed?) && operatingsystem_id_changed?
+        changed_by_dirty_tracking = respond_to?(:will_save_change_to_operatingsystem_id?) &&
+                                    will_save_change_to_operatingsystem_id?
+        changed_by_legacy_api || changed_by_dirty_tracking
+      end
+
+      def kickstart_repository_release_mismatch?
```
Comment (lfu) [inline app/models/katello/concerns/hostgroup_extensions.rb]: How about `kickstart_repository_incompatible_with_os?`?

File: app/models/katello/concerns/hostgroup_extensions.rb [hunk 6]
```diff
@@ -129,24 +123,158 @@ def add_organization_for_environment
...
+        return true if kickstart_repository_release_mismatch?
+        return true unless matching_kickstart_repository?(content_facet)
+
+        # Child hostgroups with inherited kickstart context can become stale when
+        # parent CVE/content source/OS changes while the stale repo ID remains listed.
+        ancestry.present? &&
+          (content_facet.content_view_environment_id.blank? || content_facet.content_source_id.blank?)
+      end
+
+      def operatingsystem_changed_for_recalculation?
```
Comment (lfu) [inline app/models/katello/concerns/hostgroup_extensions.rb]: How about operatingsystem_id_changing??

File: app/models/katello/concerns/hostgroup_extensions.rb [hunk 7]
```diff
@@ -129,24 +123,158 @@ def add_organization_for_environment
...
+        changed_by_legacy_api = respond_to?(:operatingsystem_id_changed?) && operatingsystem_id_changed?
+        changed_by_dirty_tracking = respond_to?(:will_save_change_to_operatingsystem_id?) &&
+                                    will_save_change_to_operatingsystem_id?
```
Comment (lfu) [inline app/models/katello/concerns/hostgroup_extensions.rb]: Why both API are called here?

File: app/models/katello/concerns/hostgroup_extensions.rb [hunk 8]
```diff
@@ -129,24 +123,158 @@ def add_organization_for_environment
...
+
+      def kickstart_repository_release_mismatch?
+        repo_release = content_facet&.kickstart_repository&.distribution_version
+        target_release = preferred_os_release_values.first
+        return false if repo_release.blank? || target_release.blank?
+
+        !(repo_release == target_release || repo_release.start_with?("#{target_release}."))
+      end
+
+      def equivalent_kickstart_repository_for_release(ks_repos)
```
Comment (lfu) [inline app/models/katello/concerns/hostgroup_extensions.rb]: `find_kickstart_repository_by_release`?

File: app/models/katello/concerns/hostgroup_extensions.rb [hunk 9]
```diff
@@ -129,24 +123,158 @@ def add_organization_for_environment
...
+
+          release_matches = matches
+          break
+        end
+        return if release_matches.blank?
+
+        equivalent_kickstart_repository_for_variant(release_matches, repos_by_id) || release_matches.first
+      end
+
+      def equivalent_kickstart_repository_for_variant(ks_repos, repos_by_id = indexed_kickstart_repositories(ks_repos))
```
Comment (lfu) [inline app/models/katello/concerns/hostgroup_extensions.rb]: `find_best_matching_kickstart_repository`

File: app/models/katello/concerns/hostgroup_extensions.rb [hunk 10]
```diff
@@ -129,24 +123,158 @@ def add_organization_for_environment
...
+          end
+          return same_product if same_product
+        end
+      end
+
+      def indexed_kickstart_repositories(ks_repos)
+        Katello::Repository.where(id: ks_repos.map { |repo| repo[:id] }).index_by(&:id)
+      end
+
+      def effective_content_facet_for_kickstart(facet)
```
Comment (lfu) [inline app/models/katello/concerns/hostgroup_extensions.rb]: `effective_content_facet` or `content_facet_with_inheritance`

File: app/models/katello/concerns/hostgroup_extensions.rb [hunk 11]
```diff
@@ -129,24 +123,158 @@ def add_organization_for_environment
...
+        inherited_content_source_id = inherited_ancestry_attribute(:content_source_id, :content_facet)
+        return facet if inherited_cvenv_id.blank? && inherited_content_source_id.blank?
+
+        facet.dup.tap do |effective_facet|
+          effective_facet.content_view_environment_id ||= inherited_cvenv_id
+          effective_facet.content_source_id ||= inherited_content_source_id
+        end
+      end
+
+      def preferred_os_release_values
```
Comment (lfu) [inline app/models/katello/concerns/hostgroup_extensions.rb]: `os_version_matching_order`

File: app/models/katello/concerns/hostgroup_extensions.rb [hunk 12]
```diff
@@ -129,24 +123,158 @@ def add_organization_for_environment
...
+        releases = []
+        major = operatingsystem&.major.to_s.presence
+        minor = operatingsystem&.minor.to_s.presence
+        releases << "#{major}.#{minor}" if major && minor
+        releases << operatingsystem.release if operatingsystem&.release.present?
+        releases.uniq
+      end
+
+      def repository_name_matches_release?(repository_name, os_release)
```
Comment (lfu) [inline app/models/katello/concerns/hostgroup_extensions.rb]: `repository_name_contains_release?`

**Comment by lfu** [comment]
PR looks good. Some comments about testing from Claude.

- **Add integration test** with minimal stubbing
- **Unit test each new private method**
- **Reduce stubbing** in existing tests
- **Shorten test names** (remove "must" prefix)
- **Add error case tests**

File: app/models/katello/concerns/hostgroup_extensions.rb (hunk 7 above)
Comment (nofaralfasi) [reply app/models/katello/concerns/hostgroup_extensions.rb]: Good point. They were originally both called to support legacy and modern dirty-tracking behavior, but I’ve updated it to avoid that: we now prefer `will_save_change_to_operatingsystem_id?` and only fall back to `operatingsystem_id_changed?` when needed.

**Review by lfu — APPROVED** [review]
LGTM 👍 
Thanks @nofaralfasi
